client/src/tests/hoc.tsx

```diff
@@ -0,0 +1,40 @@
+import React from "react";
+import { GraphQLData } from "../connection";
+import { graphql } from "../hoc";
+import { QuerySpec } from "../query";
+
+interface Result {
+  output: string;
+}
+
+interface Variables {
+  y: string;
+}
+
+const exampleQuery: QuerySpec<Result, Variables> = {
+  query: "query test {}",
+};
+
+interface Props extends GraphQLData<Result> {
+  input: string;
+}
+
+const PresentationalComponent = (props: Props) => {
+  return <div>{props.data.value.output}</div>;
+};
+
+const ConnectedComponentWithQueryAndVariables = graphql<
+  Props,
+  Result,
+  Variables
+>("query test {}", props => ({ y: props.input }))(PresentationalComponent);
+
+const instanceWithQueryAndVariables = (
+  <ConnectedComponentWithQueryAndVariables input="test" />
+);
+
+const ConnectedComponentWithSpec = graphql(exampleQuery, props => ({
```

Comment:
if you inline PresentationalComponent here, does it infer the Props type?

Reply:
sadly, i realized that it doesn't infer `props` here correctly, even not inlined.

Reply:
i think this would be more possible if the signature looked closer to what it used to, e.g. `graphql(Component, document, props => ...)`.

i tried to be too cute and it was a bad idea. see https://github.com/Microsoft/TypeScript/issues/15016#issuecomment-292108004